### PR TITLE
handle missing files when configuring

### PR DESCRIPTION
**Pull request by aerickson**
## overview

- adds a function to check that things required for configuration are present
- also pulls out configuration expansion/defaulting

### verify that test-run-manager continues to work

- in test_run_manager.py, set TESTING=True
- run test-run-manager
  - seems to operate normally

### verify that it explodes if files are missing
```
(venv) ➜  mozilla-bitbar-devicepool git:(fix_io_error) ✗ mv files files.bak
(venv) ➜  mozilla-bitbar-devicepool git:(fix_io_error) ✗  python mozilla_bitbar_devicepool/main.py run-test --update-bitbar --project-name mozilla-docker-image-test
                MainThread INFO     configure: performing checks
                MainThread ERROR    '/Users/aerickson/git/mozilla-bitbar-devicepool/files/aerickson-empty-test.zip' does not exist!
                MainThread ERROR    Configuration files seem to be missing! Please place and restart. Exiting...
(venv) ➜  mozilla-bitbar-devicepool git:(fix_io_error) ✗
```

### verify that configuration object hasn't changed

Add following to configuration.py:configure at end and compared objects.

```python
import pprint
pprint.pprint(CONFIG['projects'])
sys.exit(1)
```

Output is the same.

## original exception:

```
Jun 09 08:52:51 bitbar-devicepool-0 bash[13116]: Traceback (most recent call last):
Jun 09 08:52:51 bitbar-devicepool-0 bash[13116]:   File "/home/bitbar/mozilla-bitbar-devicepool/mozilla_bitbar_devicepool/main.py", line 170, in <module>
Jun 09 08:52:51 bitbar-devicepool-0 bash[13116]:     main()
Jun 09 08:52:51 bitbar-devicepool-0 bash[13116]:   File "/home/bitbar/mozilla-bitbar-devicepool/mozilla_bitbar_devicepool/main.py", line 167, in main
Jun 09 08:52:51 bitbar-devicepool-0 bash[13116]:     args.func(args)
Jun 09 08:52:51 bitbar-devicepool-0 bash[13116]:   File "/home/bitbar/mozilla-bitbar-devicepool/mozilla_bitbar_devicepool/main.py", line 56, in test_run_manager
Jun 09 08:52:51 bitbar-devicepool-0 bash[13116]:     configuration.configure(bitbar_configpath, filespath=args.files)
Jun 09 08:52:51 bitbar-devicepool-0 bash[13116]:   File "/home/bitbar/mozilla-bitbar-devicepool/mozilla_bitbar_devicepool/configuration.py", line 71, in configure
Jun 09 08:52:51 bitbar-devicepool-0 bash[13116]:     configure_projects()
Jun 09 08:52:51 bitbar-devicepool-0 bash[13116]:   File "/home/bitbar/mozilla-bitbar-devicepool/mozilla_bitbar_devicepool/configuration.py", line 179, in configure_projects
Jun 09 08:52:51 bitbar-devicepool-0 bash[13116]:     os.path.join(FILESPATH, file_name))
Jun 09 08:52:51 bitbar-devicepool-0 bash[13116]:   File "/home/bitbar/mozilla-bitbar-devicepool/venv/local/lib/python2.7/site-packages/testdroid/__init__.py", line 409, in upload_application_file
Jun 09 08:52:51 bitbar-devicepool-0 bash[13116]:     return self.upload(path=path, filename=filename)
Jun 09 08:52:51 bitbar-devicepool-0 bash[13116]:   File "/home/bitbar/mozilla-bitbar-devicepool/venv/local/lib/python2.7/site-packages/testdroid/__init__.py", line 237, in upload
Jun 09 08:52:51 bitbar-devicepool-0 bash[13116]:     with open(filename, 'rb') as f:
Jun 09 08:52:51 bitbar-devicepool-0 bash[13116]: IOError: [Errno 2] No such file or directory: '/home/bitbar/mozilla-bitbar-devicepool/files/aerickson-Testdroid.apk'
```

### CODE REVIEW

File: mozilla_bitbar_devicepool/configuration.py [hunk 1]
```diff
@@ -70,8 +71,13 @@ def configure(bitbar_configpath, filespath=None, update_bitbar=False):
...
     start = time.time()
-    configure_device_groups(update_bitbar=update_bitbar)
-    configure_projects(update_bitbar=update_bitbar)
+    try:
+        configure_device_groups(update_bitbar=update_bitbar)
+        configure_projects(update_bitbar=update_bitbar)
+    except IOError as e:
+        logger.error("Test files seem to be missing! Please place and restart. Exiting...")
+        logger.error("%s: %s" % (e.__class__.__name__, e.message))
+        sys.exit(1)
```
Comment (bclary) [inline mozilla_bitbar_devicepool/configuration.py]: Let's also check that you are actually seeing the missing file exception. We shouldn't say we are missing the file if we aren't positive that is the problem.

Comment (aerickson) [reply mozilla_bitbar_devicepool/configuration.py]: Via parsing the exception message or a custom exception lower in the stack?

Comment (aerickson) [reply mozilla_bitbar_devicepool/configuration.py]: I'm working on raising and handling custom exceptions.

File: mozilla_bitbar_devicepool/configuration.py [hunk 2]
```diff
@@ -191,8 +236,12 @@ def configure_projects(update_bitbar=False):
                 bitbar_file = bitbar_files[-1]
             else:
                 if update_bitbar:
-                    TESTDROID.upload_test_file(bitbar_project['id'],
-                                               os.path.join(FILESPATH, file_name))
+                    file_path = os.path.join(FILESPATH, file_name)
+                    if not os.path.exists(file_path):
+                        raise ConfigurationFileException("'%s' does not exist!" % file_path)
+                    else:
+                        print("File exists %s" % file_path)
```
Comment (bclary) [inline mozilla_bitbar_devicepool/configuration.py]: No need for the dangling else:. Do we need to print that the file exists?

File: mozilla_bitbar_devicepool/configuration.py [hunk 3]
```diff
@@ -206,8 +255,12 @@ def configure_projects(update_bitbar=False):
                 bitbar_file = bitbar_files[-1]
             else:
                 if update_bitbar:
-                    TESTDROID.upload_application_file(bitbar_project['id'],
-                                                      os.path.join(FILESPATH, file_name))
+                    file_path = os.path.join(FILESPATH, file_name)
+                    if not os.path.exists(file_path):
+                        raise ConfigurationFileException("'%s' does not exist!" % file_path)
+                    else:
+                        print("File exists %s" % file_path)
```
Comment (bclary) [inline mozilla_bitbar_devicepool/configuration.py]: ditto

File: mozilla_bitbar_devicepool/configuration.py (hunk 2 above)
Comment (aerickson) [reply mozilla_bitbar_devicepool/configuration.py]: Nope, was for debugging. Will remove.

File: mozilla_bitbar_devicepool/configuration.py (hunk 3 above)
Comment (aerickson) [reply mozilla_bitbar_devicepool/configuration.py]: Will remove this one also.

**Comment by aerickson** [comment]
> Seems ok though a little verbose. 

Parsing the exception does seem quicker. Custom exceptions feel like the cleaner solution.

For https://github.com/bclary/mozilla-bitbar-devicepool/pull/60, we don't control the code raising the exception so we can't do custom exceptions, but we can here and I think it's a bit nicer.

Updated to remove duplicate configuration expansion (aka setting defaults).

File: mozilla_bitbar_devicepool/configuration.py [hunk 4]
```diff
@@ -65,17 +75,66 @@ def configure(bitbar_configpath, filespath=None, update_bitbar=False):
...
-    with open(bitbar_configpath) as bitbar_configfile:
-        CONFIG = yaml.load(bitbar_configfile.read(), Loader=yaml.SafeLoader)
-
     logger.info('configure: starting configuration')
     start = time.time()
+
+    with open(bitbar_configpath) as bitbar_configfile:
+        CONFIG = yaml.load(bitbar_configfile.read(), Loader=yaml.SafeLoader)
+    expand_configuration()
+    if update_bitbar:
```
Comment (aerickson) [inline mozilla_bitbar_devicepool/configuration.py]: Should this explode even if we're not updating bitbar?

Comment (bclary) [reply mozilla_bitbar_devicepool/configuration.py]: I can see arguing for both sides of this. I think having it alert that there is a problem even if we aren't updating would be best. What do you think?

File: mozilla_bitbar_devicepool/configuration.py [hunk 5]
```diff
@@ -46,6 +47,15 @@
 FILESPATH = None
 CONFIG = None
 
+
+class ConfigurationException(Exception):
+    def __init__(self, message):
+        self.message = message
+
```
Comment (bclary) [inline mozilla_bitbar_devicepool/configuration.py]: another blank line so we have two blank lines between the class definitions?

File: mozilla_bitbar_devicepool/configuration.py (hunk 4 above)
Comment (bclary) [reply mozilla_bitbar_devicepool/configuration.py]: I replied in the main UI, but to comment here as well, I think so yes but am open to not doing so if you wish.

File: mozilla_bitbar_devicepool/configuration.py (hunk 5 above)
Comment (aerickson) [reply mozilla_bitbar_devicepool/configuration.py]: Done.

File: mozilla_bitbar_devicepool/configuration.py (hunk 4 above)
Comment (aerickson) [reply mozilla_bitbar_devicepool/configuration.py]: Changed to always run.

**Pull request merged**